### ISSUE TITLE
Cannot copy files with names that contain supplementary characters / surrogate pairs

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#789]
When copying files that have file names containing supplementary characters / surrogate pairs, using a `CopySpec` or from the resources folder when using the `java` plugin, Gradle is not capable of copying them, reporting that the files do not exist.

### Expected Behavior

Copy should work.

### Current Behavior

Copy fails:

```
Caused by: org.gradle.api.GradleException: Could not copy file '/Users/cgal0002/workspace/github.com/cjbooms/japanese-characters-maven/src/test/resources/supplementary/アンドリューは本当に凄いですawesomeだと思います𩸽.jpg' to '/Users/cgal0002/workspace/github.ccharacters-maven/build/resources/test/supplementary/アンドリューは本当に凄いですawesomeだと思います𩸽.jpg'.
        at org.gradle.api.internal.file.AbstractFileTreeElement.copyTo(AbstractFileTreeElement.java:79)
        ... 68 more
Caused by: org.gradle.internal.nativeintegration.filesystem.FileException: Could not get file mode for '/Users/cgal0002/workspace/github.com/cjbooms/japanese-characters-maven/src/test/resources/supplementary/アンドリューは本当に凄いですawesomeだと思います𩸽.jpg'.
        at org.gradle.internal.nativeintegration.filesystem.services.GenericFileSystem.getUnixMode(GenericFileSystem.java:72)
        ... 105 more
Caused by: net.rubygrapefruit.platform.NativeException: Could not get UNIX mode on /Users/cgal0002/workspace/github.com/cjbooms/japanese-characters-maven/src/test/resources/supplementary/アンドリューは本当に凄いですawesomeだと思います𩸽.jpg: file does not exist.
        at net.rubygrapefruit.platform.internal.DefaultPosixFiles.getMode(DefaultPosixFiles.java:48)
        at org.gradle.internal.nativeintegration.filesystem.services.NativePlatformBackedStat.getUnixMode(NativePlatformBackedStat.java:34)
        at org.gradle.internal.nativeintegration.filesystem.services.GenericFileSystem.getUnixMode(GenericFileSystem.java:70)
        ... 107 more
```

### Context

This issue has been reported on the forum: https://discuss.gradle.org/t/build-fails-to-copy-resources-with-special-characters-in-file-name/20294

### Steps to Reproduce (for bugs)

A reproducer project is available: https://github.com/cjbooms/japanese-filenames

### Your Environment

```
Gradle Version: 3.1
Operating System: MAC OS X El Capitan 
JVM Version:
java version "1.8.0_51"
Java(TM) SE Runtime Environment (build 1.8.0_51-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.51-b03, mixed mode)
```

Fails with Gradle 2.x, 3.x.

**Comment by eskatos** [comment]
Reproduced on macOS, Linux and Windows.
Root cause and stack trace are the same.

**Comment by lhotari** [comment]
This might be related to  [GRADLE-2181](https://issues.gradle.org/browse/GRADLE-2181) . These links might contain useful background information: http://happygiraffe.net/blog/2009/09/24/java-platform-encoding/ , http://stackoverflow.com/questions/11343828/file-encoding-has-no-effect-lc-all-environment-variable-does-it . TL;DR The `file.encoding` doesn't control file name encoding. There is a hidden system property `sun.jnu.encoding`, but that shouldn't be manipulated directly.
Search the Gradle source code with GRADLE-2181 to find some tests related to that issue.

**Comment by hotchemi** [comment]
Anyone working on it? Can I grab?

**Comment by lacasseio** [comment]
Thanks @hotchemi for your interest in this issue. No one is assigned to this issue. I see no problem for you to take the lead on this. Don't hesitate to ask questions if you feel stuck.

**Comment by hotchemi** [comment]
@lacasseio Thx, I'll try!

**Comment by hotchemi** [comment]
@lacasseio Sorry for the late. I digged into the problem and being little bit stuck...so let me share current status.

#### Cause

I followed the stack trace and it turned out the original exception is caused inside of nativeplatform's `DefaultPosixFiles.java` and `posix.cpp`.

- [DefaultPosixFiles.java](https://github.com/adammurdoch/native-platform/blob/674992d4692355ce435167228a90e35c39112fe9/src/main/java/net/rubygrapefruit/platform/internal/DefaultPosixFiles.java#L68)
- [posix.cpp](https://github.com/adammurdoch/native-platform/blob/0df1fb6d563f37dcec355da9edd2fd3b1d3d46c1/src/main/cpp/posix.cpp#L136)

**It seems we gotta give unicode escaped string to this library to execute stat**(here's a [reproducer test](https://github.com/hotchemi/native-platform/pull/1/files)).

#### Workaround?

Therefore, I tried to add a [simple test case](https://github.com/hotchemi/gradle/pull/1/files) that adding convert/reconvert file name procedures before and after copy task. And it works well...I know it's kinda hucky though.
The reason I add this procedures is as  @lhotari said global encoding option shouldn't be manipulated directly.

#### Things to discuss?

So I'm just thinking two ways to fix the problem...

- Adding a convert/reconvert file name procedures if file name contains surrogate pairs or so as I added above inside of gradle.
- Fix adammurdoch/native-platform to allow surrogate pairs(sorry I'm not sure C++ and this library yet🙇)

Say, can you give me your thoughts or idea to fix the problem cleverly? This is the first time I dig into Gradle source code so I know you know a lot of things that I don't know. Thx😃

**Comment by lacasseio** [comment]
Thanks for looking into this @hotchemi. I think the preferred solution should be done in [`adammurdoch/native-platform`](https://github.com/adammurdoch/native-platform). I suspect [the `java_to_char` call](https://github.com/adammurdoch/native-platform/blob/0df1fb6d563f37dcec355da9edd2fd3b1d3d46c1/src/main/cpp/posix.cpp#L119) should be changed to `java_to_wchar`. Unfortunately, `stat` and `lstat` doesn't support `wchar_t` path in Posix. You will need to find a way to achieve this.

@eskatos Would you be able to comment on how handling `wchar` in native-platform affect the performance?

**Comment by hotchemi** [comment]
@lacasseio Thx for your quick response. I tried to address with `wcstombs` on my local but `java_to_wchar` is only defined in `win.cpp`...I'd like to hear @eskatos's thought, sorry to bother you🙇

**Comment by eskatos** [comment]
On the native side of JNI you get Java Strings as `jstring`s. These are [Modified UTF-8](https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8) strings.

`char` and `wchar` arrays can be used to convey differently encoded strings. How they should be encoded depends on the targeted native method and what it expects.

`java_to_char` in `native-platform` already does a great job converting `jstrings` but maybe it does it in a way that is not expected by `stat` and `lstat`.

A good first step would be to add a test case to `native-platform` that exhibits the issue.

**Comment by OleKsimov** [comment]
Still an active issue :(

**Comment by GrayYoga** [comment]
In 2021, the problem is still relevant for cyrillic filenames
```
> Task :core:processResources FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':core:processResources'.
> Could not copy file '/builds/[project]/external-autotests/core/src/main/resources/clientfiles/?????????????????????? ??????????????.txt' to '/builds/[project]/external-autotests/core/build/resources/main/clientfiles/?????????????????????? ??????????????.txt'.
   > Problems opening file input stream for file: /builds/[project]/external-autotests/core/src/main/resources/clientfiles/?????????????????????? ??????????????.txt

```

**Comment by ov7a** [comment]
Sorry for the late reply.

This was fixed in 9.3.1 by
* https://github.com/gradle/gradle/pull/36431
